File: frontend/src/api/http-execution.ts
import { resolveBackendConnection } from "./backend-connection";
import { ApiClient, ApiClientError } from "./client";

export type HttpExecutionStatus =
  | "pending"
  | "queued"
  | "running"
  | "passed"
  | "failed"
  | "error"
  | "cancelled"
  | "timeout";
export type HttpMethod = "DELETE" | "GET" | "HEAD" | "OPTIONS" | "PATCH" | "POST" | "PUT";
export type HttpAssertionKind =
  | "status_code"
  | "header_equals"
  | "body_contains"
  | "json_path_equals";
export interface HttpAssertion {
  kind: HttpAssertionKind;
  target: string | null;
  expected: string;
}
export interface HttpAssertionResult extends HttpAssertion {
  actual: string | null;
  passed: boolean;
  message: string;
}
export interface HttpExecutionEvent {
  id: string;
  ordinal: number;
  level: "info" | "warning" | "error";
  code: string;
  message: string;
  attempt: number | null;
  created_at: string;
}

export interface HttpEnvironment {
  id: string;
  workspace_id: string;
  name: string;
  base_url: string;
  variables: Record<string, string>;
  secret_names: string[];
  created_at: string;
  updated_at: string;
}

export interface HttpEnvironmentInput {
  name: string;
  base_url: string;
  variables: Record<string, string>;
}

export interface HttpExecutionStartInput {
  environment_id: string;
  method: HttpMethod;
  path: string;
  headers: Record<string, string>;
  body: string | null;
  timeout_seconds: number;
  max_attempts: number;
  assertions: HttpAssertion[];
}

export interface HttpExecution {
  id: string;
  workspace_id: string;
  environment_id: string | null;
  environment_name: string;
  method: HttpMethod;
  base_url: string;
  path_template: string;
  headers_template: Record<string, string>;
  body_template: string | null;
  timeout_seconds: number;
  max_attempts: number;
  assertions: HttpAssertion[];
  assertion_results: HttpAssertionResult[];
  events: HttpExecutionEvent[];
  status: HttpExecutionStatus;
  progress: number;
  response_status_code: number | null;
  response_headers: Record<string, string>;
  response_body: string | null;
  response_body_encoding: "text" | "base64" | null;
  response_size_bytes: number | null;
  duration_ms: number | null;
  error_code: string | null;
  error_message: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
}

const methods = new Set<HttpMethod>(["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]);
const statuses = new Set<HttpExecutionStatus>([
  "pending",
  "queued",
  "running",
  "passed",
  "failed",
  "error",
  "cancelled",
  "timeout",
]);

function stringMap(value: unknown): value is Record<string, string> {
  return (
    typeof value === "object" &&
    value !== null &&
    !Array.isArray(value) &&
    Object.entries(value).every(([key, item]) => typeof key === "string" && typeof item === "string")
  );
}

function nullableString(value: unknown): boolean {
  return value === null || typeof value === "string";
}

function isEnvironment(value: unknown): value is HttpEnvironment {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<HttpEnvironment>;
  return (
    typeof item.id === "string" &&
    typeof item.workspace_id === "string" &&
    typeof item.name === "string" &&
    typeof item.base_url === "string" &&
    stringMap(item.variables) &&
    Array.isArray(item.secret_names) &&
    item.secret_names.every((name) => typeof name === "string") &&
    typeof item.created_at === "string" &&
    typeof item.updated_at === "string"
  );
}

function isExecution(value: unknown): value is HttpExecution {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<HttpExecution>;
  return (
    typeof item.id === "string" &&
    typeof item.workspace_id === "string" &&
    nullableString(item.environment_id) &&
    typeof item.environment_name === "string" &&
    methods.has(item.method as HttpMethod) &&
    typeof item.base_url === "string" &&
    typeof item.path_template === "string" &&
    stringMap(item.headers_template) &&
    nullableString(item.body_template) &&
    typeof item.timeout_seconds === "number" &&
    typeof item.max_attempts === "number" &&
    Array.isArray(item.assertions) &&
    item.assertions.every(isAssertion) &&
    Array.isArray(item.assertion_results) &&
    item.assertion_results.every(isAssertionResult) &&
    Array.isArray(item.events) &&
    item.events.every(isEvent) &&
    statuses.has(item.status as HttpExecutionStatus) &&
    typeof item.progress === "number" &&
    (item.response_status_code === null || typeof item.response_status_code === "number") &&
    stringMap(item.response_headers) &&
    nullableString(item.response_body) &&
    (item.response_body_encoding === null ||
      item.response_body_encoding === "text" ||
      item.response_body_encoding === "base64") &&
    (item.response_size_bytes === null || typeof item.response_size_bytes === "number") &&
    (item.duration_ms === null || typeof item.duration_ms === "number") &&
    nullableString(item.error_code) &&
    nullableString(item.error_message) &&
    typeof item.created_at === "string" &&
    nullableString(item.started_at) &&
    nullableString(item.finished_at)
  );
}

function isAssertion(value: unknown): value is HttpAssertion {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<HttpAssertion>;
  return (
    ["status_code", "header_equals", "body_contains", "json_path_equals"].includes(
      item.kind ?? "",
    ) &&
    nullableString(item.target) &&
    typeof item.expected === "string"
  );
}

function isAssertionResult(value: unknown): value is HttpAssertionResult {
  if (!isAssertion(value)) return false;
  const item = value as Partial<HttpAssertionResult>;
  return (
    nullableString(item.actual) &&
    typeof item.passed === "boolean" &&
    typeof item.message === "string"
  );
}

function isEvent(value: unknown): value is HttpExecutionEvent {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<HttpExecutionEvent>;
  return (
    typeof item.id === "string" &&
    typeof item.ordinal === "number" &&
    ["info", "warning", "error"].includes(item.level ?? "") &&
    typeof item.code === "string" &&
    typeof item.message === "string" &&
    (item.attempt === null || typeof item.attempt === "number") &&
    typeof item.created_at === "string"
  );
}

function validateEnvironment(value: unknown): HttpEnvironment {
  if (!isEnvironment(value)) {
    throw new ApiClientError("后端 HTTP 环境响应格式不正确。", { code: "INVALID_RESPONSE" });
  }
  return value;
}

function validateEnvironments(value: unknown): HttpEnvironment[] {
  if (!Array.isArray(value) || !value.every(isEnvironment)) {
    throw new ApiClientError("后端 HTTP 环境列表格式不正确。", { code: "INVALID_RESPONSE" });
  }
  return value;
}

function validateExecution(value: unknown): HttpExecution {
  if (!isExecution(value)) {
    throw new ApiClientError("后端 HTTP 执行响应格式不正确。", { code: "INVALID_RESPONSE" });
  }
  return value;
}

function validateExecutions(value: unknown): HttpExecution[] {
  if (!Array.isArray(value) || !value.every(isExecution)) {
    throw new ApiClientError("后端 HTTP 执行列表格式不正确。", { code: "INVALID_RESPONSE" });
  }
  return value;
}

async function client(): Promise<ApiClient> {
  const connection = await resolveBackendConnection();
  return new ApiClient(connection.baseUrl, connection.token);
}

export async function listHttpEnvironments(workspaceId: string): Promise<HttpEnvironment[]> {
  return validateEnvironments(
    await (await client()).get<unknown>(`/api/workspaces/${workspaceId}/http-environments`),
  );
}

export async function createHttpEnvironment(
  workspaceId: string,
  input: HttpEnvironmentInput,
): Promise<HttpEnvironment> {
  return validateEnvironment(
    await (await client()).post<unknown>(`/api/workspaces/${workspaceId}/http-environments`, input),
  );
}

export async function updateHttpEnvironment(
  workspaceId: string,
  environmentId: string,
  input: HttpEnvironmentInput,
): Promise<HttpEnvironment> {
  return validateEnvironment(
    await (await client()).put<unknown>(
      `/api/workspaces/${workspaceId}/http-environments/${environmentId}`,
      input,
    ),
  );
}

export async function deleteHttpEnvironment(
  workspaceId: string,
  environmentId: string,
): Promise<void> {
  await (await client()).delete<void>(
    `/api/workspaces/${workspaceId}/http-environments/${environmentId}`,
  );
}

export async function setHttpSecret(
  workspaceId: string,
  environmentId: string,
  name: string,
  secret: string,
): Promise<HttpEnvironment> {
  return validateEnvironment(
    await (await client()).put<unknown>(
      `/api/workspaces/${workspaceId}/http-environments/${environmentId}/secrets/${encodeURIComponent(name)}`,
      { secret },
    ),
  );
}

export async function deleteHttpSecret(
  workspaceId: string,
  environmentId: string,
  name: string,
): Promise<HttpEnvironment> {
  return validateEnvironment(
    await (await client()).delete<unknown>(
      `/api/workspaces/${workspaceId}/http-environments/${environmentId}/secrets/${encodeURIComponent(name)}`,
    ),
  );
}

export async function startHttpExecution(
  workspaceId: string,
  input: HttpExecutionStartInput,
): Promise<HttpExecution> {
  return validateExecution(
    await (await client()).post<unknown>(`/api/workspaces/${workspaceId}/http-executions`, input),
  );
}

export async function listHttpExecutions(workspaceId: string): Promise<HttpExecution[]> {
  return validateExecutions(
    await (await client()).get<unknown>(`/api/workspaces/${workspaceId}/http-executions`),
  );
}

export async function getHttpExecution(
  workspaceId: string,
  runId: string,
): Promise<HttpExecution> {
  return validateExecution(
    await (await client()).get<unknown>(`/api/workspaces/${workspaceId}/http-executions/${runId}`),
  );
}

export async function cancelHttpExecution(
  workspaceId: string,
  runId: string,
): Promise<HttpExecution> {
  return validateExecution(
    await (await client()).post<unknown>(
      `/api/workspaces/${workspaceId}/http-executions/${runId}/cancel`,
    ),
  );
}

export async function rerunHttpExecution(
  workspaceId: string,
  runId: string,
): Promise<HttpExecution> {
  return validateExecution(
    await (await client()).post<unknown>(
      `/api/workspaces/${workspaceId}/http-executions/${runId}/rerun`,
    ),
  );
}

File: backend/app/domain/http_execution.py
from __future__ import annotations

import json
import re
from collections.abc import Mapping
from dataclasses import dataclass
from datetime import datetime
from typing import Literal
from urllib.parse import urlsplit, urlunsplit

HttpExecutionStatus = Literal[
    "pending", "queued", "running", "passed", "failed", "error", "cancelled", "timeout"
]
HttpMethod = Literal["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
HttpAssertionKind = Literal["status_code", "header_equals", "body_contains", "json_path_equals"]
HttpEventLevel = Literal["info", "warning", "error"]
TERMINAL_HTTP_EXECUTION_STATUSES: frozenset[HttpExecutionStatus] = frozenset(
    {"passed", "failed", "error", "cancelled", "timeout"}
)
HTTP_METHODS: frozenset[str] = frozenset(
    {"DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"}
)
VARIABLE_NAME_PATTERN = re.compile(r"^[A-Z][A-Z0-9_]{0,63}$")
TEMPLATE_REFERENCE_PATTERN = re.compile(r"{{(secret\.)?([A-Z][A-Z0-9_]{0,63})}}")
JSON_PATH_PATTERN = re.compile(r"^\$(?:\.(?:[A-Za-z_][A-Za-z0-9_-]*|\d+))+$")


class HttpTemplateError(ValueError):
    pass


def validate_variable_name(name: str) -> str:
    if VARIABLE_NAME_PATTERN.fullmatch(name) is None:
        raise ValueError("invalid variable name")
    return name


def validate_and_normalize_base_url(value: str) -> str:
    candidate = value.strip()
    parsed = urlsplit(candidate)
    if (
        parsed.scheme not in {"http", "https"}
        or not parsed.hostname
        or parsed.username is not None
        or parsed.password is not None
        or parsed.query
        or parsed.fragment
    ):
        raise ValueError("unsafe HTTP base URL")
    try:
        _ = parsed.port
    except ValueError as exception:
        raise ValueError("invalid HTTP base URL port") from exception
    path = parsed.path.rstrip("/")
    return urlunsplit((parsed.scheme, parsed.netloc, path, "", ""))


def resolve_template(
    template: str,
    *,
    variables: Mapping[str, str],
    secrets: Mapping[str, str],
) -> str:
    cursor = 0
    resolved: list[str] = []
    for match in TEMPLATE_REFERENCE_PATTERN.finditer(template):
        prefix = template[cursor : match.start()]
        if "{{" in prefix or "}}" in prefix:
            raise HttpTemplateError("malformed template reference")
        resolved.append(prefix)
        secret_prefix, name = match.groups()
        source = secrets if secret_prefix else variables
        if name not in source:
            raise HttpTemplateError("template reference is not configured")
        resolved.append(source[name])
        cursor = match.end()
    suffix = template[cursor:]
    if "{{" in suffix or "}}" in suffix:
        raise HttpTemplateError("malformed template reference")
    resolved.append(suffix)
    return "".join(resolved)


def redact_secrets(value: str, secrets: list[str] | tuple[str, ...]) -> str:
    redacted = value
    for secret in sorted((item for item in secrets if item), key=len, reverse=True):
        redacted = redacted.replace(secret, "***")
    return redacted


@dataclass(frozen=True, slots=True)
class HttpAssertion:
    kind: HttpAssertionKind
    target: str | None
    expected: str

    def validate(self) -> HttpAssertion:
        if not self.expected or len(self.expected) > 4000:
            raise ValueError("invalid assertion expected value")
        if self.kind == "status_code":
            if self.target is not None or not self.expected.isdigit():
                raise ValueError("invalid status assertion")
            expected_status = int(self.expected)
            if not 100 <= expected_status <= 599:
                raise ValueError("invalid expected status")
        elif self.kind == "header_equals":
            if self.target is None or not self.target or len(self.target) > 128:
                raise ValueError("invalid header assertion")
        elif self.kind == "body_contains":
            if self.target is not None:
                raise ValueError("invalid body assertion")
        elif self.kind == "json_path_equals":
            if self.target is None or JSON_PATH_PATTERN.fullmatch(self.target) is None:
                raise ValueError("invalid JSON path assertion")
            expected = json.loads(self.expected)
            if isinstance(expected, (dict, list)):
                raise ValueError("JSON assertion expected value must be scalar")
        else:
            raise ValueError("unsupported assertion kind")
        return self


@dataclass(frozen=True, slots=True)
class HttpAssertionResult:
    kind: HttpAssertionKind
    target: str | None
    expected: str
    actual: str | None
    passed: bool
    message: str


@dataclass(frozen=True, slots=True)
class HttpExecutionEvent:
    id: str
    ordinal: int
    level: HttpEventLevel
    code: str
    message: str
    attempt: int | None
    created_at: datetime


def _display(value: object) -> str:
    rendered = json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    return rendered if len(rendered) <= 500 else f"{rendered[:497]}..."


def _json_path_value(document: object, path: str) -> tuple[bool, object]:
    current = document
    for segment in path.split(".")[1:]:
        if isinstance(current, dict) and segment in current:
            current = current[segment]
        elif isinstance(current, list) and segment.isdigit() and int(segment) < len(current):
            current = current[int(segment)]
        else:
            return False, None
    return True, current


def evaluate_http_assertions(
    assertions: tuple[HttpAssertion, ...],
    *,
    status_code: int,
    headers: Mapping[str, str],
    body: str,
    body_encoding: Literal["text", "base64"],
) -> tuple[HttpAssertionResult, ...]:
    normalized_headers = {name.casefold(): value for name, value in headers.items()}
    parsed_body: object | None = None
    parsed_body_ready = False
    results: list[HttpAssertionResult] = []
    for assertion in assertions:
        assertion.validate()
        actual: str | None = None
        passed = False
        message = "断言未通过。"
        if assertion.kind == "status_code":
            actual = str(status_code)
            passed = actual == assertion.expected
            message = "HTTP 状态码符合预期。" if passed else "HTTP 状态码与预期不一致。"
        elif assertion.kind == "header_equals":
            actual = normalized_headers.get((assertion.target or "").casefold())
            passed = actual == assertion.expected
            message = "响应头符合预期。" if passed else "响应头缺失或与预期不一致。"
        elif assertion.kind == "body_contains":
            if body_encoding == "text":
                passed = assertion.expected in body
                message = "响应正文包含预期文本。" if passed else "响应正文未包含预期文本。"
            else:
                message = "二进制响应不支持正文包含断言。"
        elif assertion.kind == "json_path_equals":
            if body_encoding != "text":
                message = "二进制响应不支持 JSON 路径断言。"
            else:
                if not parsed_body_ready:
                    try:
                        parsed_body = json.loads(body)
                    except json.JSONDecodeError:
                        parsed_body = None
                    parsed_body_ready = True
                found, value = _json_path_value(parsed_body, assertion.target or "")
                if found:
                    actual = _display(value)
                    expected_value = json.loads(assertion.expected)
                    passed = value == expected_value and type(value) is type(expected_value)
                message = "JSON 路径值符合预期。" if passed else "JSON 路径缺失或值不一致。"
        results.append(
            HttpAssertionResult(
                assertion.kind,
                assertion.target,
                assertion.expected,
                actual,
                passed,
                message,
            )
        )
    return tuple(results)


@dataclass(frozen=True, slots=True)
class HttpEnvironment:
    id: str
    workspace_id: str
    name: str
    base_url: str
    variables: dict[str, str]
    secret_names: tuple[str, ...]
    created_at: datetime
    updated_at: datetime


@dataclass(frozen=True, slots=True)
class HttpEnvironmentInput:
    name: str
    base_url: str
    variables: dict[str, str]


@dataclass(frozen=True, slots=True)
class HttpExecutionStartInput:
    environment_id: str
    method: HttpMethod
    path: str
    headers: dict[str, str]
    body: str | None
    timeout_seconds: int
    max_attempts: int = 1
    assertions: tuple[HttpAssertion, ...] = ()


@dataclass(frozen=True, slots=True)
class HttpExecution:
    id: str
    workspace_id: str
    environment_id: str | None
    environment_name: str
    method: HttpMethod
    base_url: str
    path_template: str
    headers_template: dict[str, str]
    body_template: str | None
    timeout_seconds: int
    status: HttpExecutionStatus
    progress: int
    pid: int | None
    response_status_code: int | None
    response_headers: dict[str, str]
    response_body: str | None
    response_body_encoding: Literal["text", "base64"] | None
    response_size_bytes: int | None
    duration_ms: int | None
    error_code: str | None
    error_message: str | None
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    max_attempts: int = 1
    assertions: tuple[HttpAssertion, ...] = ()
    assertion_results: tuple[HttpAssertionResult, ...] = ()
    events: tuple[HttpExecutionEvent, ...] = ()

    @property
    def can_cancel(self) -> bool:
        return self.status not in TERMINAL_HTTP_EXECUTION_STATUSES


@dataclass(frozen=True, slots=True)
class HttpExecutionTaskRequest:
    run_id: str


@dataclass(frozen=True, slots=True)
class HttpExecutionInput:
    run_id: str
    base_url: str
    variables: dict[str, str]
    secret_names: tuple[str, ...]
    method: HttpMethod
    path_template: str
    headers_template: dict[str, str]
    body_template: str | None
    timeout_seconds: int
    max_attempts: int = 1
    assertions: tuple[HttpAssertion, ...] = ()


@dataclass(frozen=True, slots=True)
class HttpExecutionResult:
    status_code: int
    headers: dict[str, str]
    body: str
    body_encoding: Literal["text", "base64"]
    size_bytes: int
    duration_ms: int
